src/markov/ddp.jl

```diff
@@ -0,0 +1,394 @@
+"""
+function solve{Algo<:DDPAlgorithm}(ddp::DiscreteDP, method::Type{Algo}=VFI;
+                                   max_iter::Int=250, epsilon::Float64=1e-3,
+                                   k::Int=20)
+    ddpr = DPSolveResult{Algo}(ddp)
+    _solve!(ddp, ddpr, max_iter, epsilon, k)
+    ddpr.mc = MarkovChain(ddp, ddpr)
+    ddpr
+end
+
+"""
+Impliments Value Iteration
+NOTE: See `solve` for further details
+"""
+function _solve!(ddp::DiscreteDP, ddpr::DPSolveResult{VFI}, max_iter::Int64,
+               epsilon::Float64, k::Int)
+    tol = epsilon * (1-ddp.beta) / (2*ddp.beta)
+
+    for i in 1:max_iter
+        # updates Tv in place
+        bellman_operator!(ddp, ddpr)
+
+        # compute error and update the v inside ddpr
+        err = maxabs(ddpr.Tv .- ddpr.v)
+        copy!(ddpr.v, ddpr.Tv)
+        ddpr.num_iter += 1
+
+        if err < tol
+            break
+        end
+    end
+
+    ddpr
+end
+
+"""
+Policy Function Iteration
+
+NOTE: The epsilon is ignored in this method. It is only here so dispatch can
+      go from `solve(::DiscreteDP, ::Type{Algo})` to any of the algorithms.
+      See `solve` for further details
+"""
+function _solve!(ddp::DiscreteDP, ddpr::DPSolveResult{PFI}, max_iter::Int,
+                epsilon::Float64, k::Int)
+    old_sigma = copy(ddpr.sigma)
+
+    for i in 1:max_iter
+       ddpr.v = evaluate_policy(ddp, ddpr)
+       compute_greedy!(ddp, ddpr)
+
+       if all(old_sigma .== ddpr.sigma)
+           break
+       end
+       copy!(old_sigma, ddpr.sigma)
+       ddpr.num_iter += 1
+    end
+
+    ddpr
+end
+
+span(x::Vector) = maximum(x) - minimum(x)
+midrange(x::Vector) = mean(extrema(x))
+
+"""
+Policy Function Iteration
+
+NOTE: The epsilon is ignored in this method. It is only here so dispatch can
+      go from `solve(::DiscreteDP, ::Type{Algo})` to any of the algorithms.
+      See `solve` for further details
+"""
+function _solve!(ddp::DiscreteDP, ddpr::DPSolveResult{MPFI}, max_iter::Int,
+                 epsilon::Float64, k::Int)
+    beta = ddp.beta
+    fill!(ddpr.v, minimum(ddp.R[ddp.R .> -Inf]) / (1.0 - beta))
+    old_sigma = copy(ddpr.sigma)
+
+    tol = beta > 0 ? epsilon * (1-beta) / beta : Inf
+
+    for i in 1:max_iter
+        bellman_operator!(ddp, ddpr)  # updates Tv, sigma inplace
+        dif = ddpr.Tv - ddpr.v
+
+        # check convergence
+        if span(dif) < tol
+            ddpr.v = ddpr.Tv + midrange(dif) * beta / (1-beta)
+            break
+        end
+
+        # now update v to use the output of the bellman step when entering
+        # policy loop
+        copy!(ddpr.v, ddpr.Tv)
+
+        # now do k iterations of policy iteration
+        for i in 1:k
+            R_sigma, Q_sigma = RQ_sigma(ddp, ddpr)
+            ddpr.Tv = R_sigma + beta * Q_sigma * ddpr.v
+
+            # compute error and update the v inside ddpr
+            err = maxabs(ddpr.Tv .- ddpr.v)
+            copy!(ddpr.v, ddpr.Tv)
+
+            if err < tol
```

Comment:
This part is not there in the Python version.

There is a tradeoff between the benefit of avoiding "unnecessary" iterations by breaking when `err < tol` and the cost of computing `err = maxabs(ddpr.Tv .- ddpr.v)` in every iteration, and this depends on the value of `tol`. In general, there is no reason that the tolerance here should be the same value as the tolerance used in the main loop for  `span(dif) < tol`. If we keep this part, maybe better to give the option of choose an independent tolerance value?